app/controllers/incomeSources/add/AddBusinessNameController.scala

```diff
@@ -151,30 +152,39 @@ class AddBusinessNameController @Inject()(authenticate: AuthenticationPredicate,
   }
 
   def handleSubmitRequest(isAgent: Boolean, isChange: Boolean)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
-    val errorHandler = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
     val (backUrlLocal, submitActionLocal, redirectLocal) = (isAgent, isChange) match {
       case (false, false) => (backUrl, submitAction, redirect)
       case (true, false) => (backUrlAgent, submitActionAgent, redirectAgent)
       case (false, true) => (checkDetailsBackUrl, submitChangeAction, checkDetailsRedirect)
       case (true, true) => (checkDetailsBackUrlAgent, submitChangeActionAgent, checkDetailsRedirectAgent)
     }
+    sessionService.get(SessionKeys.businessTrade).flatMap {
+      case Right(businessTradeName) =>
+        BusinessNameForm.checkBusinessNameWithTradeName(BusinessNameForm.form.bindFromRequest(), businessTradeName).fold(
+          formWithErrors =>
+            Future.successful(
+              Ok(addBusinessView(formWithErrors,
+                isAgent,
+                submitActionLocal,
+                backUrlLocal,
+                useFallbackLink = true))),
+
+          formData => {
+            val redirect = Redirect(redirectLocal)
+            sessionService.set(SessionKeys.businessName, formData.name, redirect).flatMap {
+              case Right(result) => Future.successful(result)
+              case Left(exception) => Future.failed(exception)
+            }
+          }
+        )
 
-    BusinessNameForm.form.bindFromRequest().fold(
-      formWithErrors => {
-        Future {
-          Ok(addBusinessView(formWithErrors, isAgent, submitActionLocal, backUrlLocal, useFallbackLink = true))
-        }
-      },
-      formData => {
-        val redirect = Redirect(redirectLocal)
-        sessionService.set(SessionKeys.businessName, formData.name, redirect).map {
-          case Right(result) => result
-          case Left(error) =>
-            Logger("application").error(s"[AddBusinessNameController][handleRequest] $error")
-            errorHandler.showInternalServerError()
-        }
-      }
-    )
+      case Left(exception) => Future.failed(exception)
+    }
+  }.recover {
+    case exception =>
+      val errorHandler = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+      Logger("application").error(s"[AddBusinessNameController][handleSubmitRequest] ${exception.getMessage}")
+      errorHandler.showInternalServerError()
```

Comment:
I'd suggest handling this exception in a more functional way, will send you an example on Slack